hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java

```diff
@@ -205,17 +216,27 @@ private void initMetadataReader() {
    */
   private void enablePartitions() {
     final HoodieMetadataConfig metadataConfig = dataWriteConfig.getMetadataConfig();
-    if (dataWriteConfig.isMetadataTableEnabled() || dataMetaClient.getTableConfig().isMetadataPartitionAvailable(MetadataPartitionType.FILES)) {
-      this.enabledPartitionTypes.add(MetadataPartitionType.FILES);
+    if (dataWriteConfig.isMetadataTableEnabled() || dataMetaClient.getTableConfig().isMetadataPartitionAvailable(FILES)) {
+      this.enabledPartitionTypes.add(FILES);
     }
-    if (metadataConfig.isBloomFilterIndexEnabled() || dataMetaClient.getTableConfig().isMetadataPartitionAvailable(MetadataPartitionType.BLOOM_FILTERS)) {
-      this.enabledPartitionTypes.add(MetadataPartitionType.BLOOM_FILTERS);
+    if (metadataConfig.isBloomFilterIndexEnabled() || dataMetaClient.getTableConfig().isMetadataPartitionAvailable(BLOOM_FILTERS)) {
+      this.enabledPartitionTypes.add(BLOOM_FILTERS);
     }
-    if (metadataConfig.isColumnStatsIndexEnabled() || dataMetaClient.getTableConfig().isMetadataPartitionAvailable(MetadataPartitionType.COLUMN_STATS)) {
-      this.enabledPartitionTypes.add(MetadataPartitionType.COLUMN_STATS);
+    if (metadataConfig.isColumnStatsIndexEnabled() || dataMetaClient.getTableConfig().isMetadataPartitionAvailable(COLUMN_STATS)) {
+      this.enabledPartitionTypes.add(COLUMN_STATS);
     }
-    if (dataWriteConfig.isRecordIndexEnabled() || dataMetaClient.getTableConfig().isMetadataPartitionAvailable(MetadataPartitionType.RECORD_INDEX)) {
-      this.enabledPartitionTypes.add(MetadataPartitionType.RECORD_INDEX);
+    if (dataWriteConfig.isRecordIndexEnabled() || dataMetaClient.getTableConfig().isMetadataPartitionAvailable(RECORD_INDEX)) {
+      this.enabledPartitionTypes.add(RECORD_INDEX);
+    }
+    Option<HoodieFunctionalIndexConfig> functionalIndexConfig = dataMetaClient.getFunctionalIndexConfig();
+    if (functionalIndexConfig.isPresent()) {
+      this.enabledPartitionTypes.add(FUNCTIONAL_INDEX);
```

Comment:
I'm trying to understand how `FUNCTIONAL_INDEX` is enabled.  Based on the code, `CREATE INDEX` can add a new functional index partition.  What about the write configs?  Are users allowed to create a functional index through write configs?

Reply:
create index will add a new partition and also the table config for available metadata partitions will be updated. That way, readers can use the index as well as subsequent writes can update the index.